electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +263,107 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
+
+  def contents(self):
+    super().contents()
+
+    control_voltage = self.control.link().voltage.hull(RangeExpr.ZERO,)
+    pwr_voltage = self.pwr_out.link().voltage.hull(RangeExpr.ZERO,)
+    pwr_current = self.pwr_out.link().current_drawn.hull(RangeExpr.ZERO,)
+
+    self.pull_res = self.Block(Resistor(
+      resistance=self.pull_resistance
+    ))
+    self.pwr_fet = self.Block(Fet.PFet(
+      drain_voltage=pwr_voltage,
+      drain_current=pwr_current,
+      gate_voltage=(control_voltage.upper(), control_voltage.upper()),  # TODO this ignores the diode drop
+    ))
+
+    self.amp_res = self.Block(Resistor(
+      resistance=self.amp_resistance
+    ))
+    self.amp_fet = self.Block(Fet.NFet(
+      drain_voltage=control_voltage,
+      drain_current=RangeExpr.ZERO,  # effectively no current
+      gate_voltage=(self.control.link().output_thresholds.upper(), self.control.link().voltage.upper())
+    ))
+
+    self.ctl_diode = self.Block(Diode(
+      reverse_voltage=control_voltage,
+      current=RangeExpr.ZERO,  # effectively no current
+      voltage_drop=self.diode_drop,
+      reverse_recovery_time=RangeExpr.ALL
+    ))
+
+    self.btn_diode = self.Block(Diode(
+      reverse_voltage=control_voltage,
+      current=RangeExpr.ZERO,  # effectively no current
+      voltage_drop=self.diode_drop,
+      reverse_recovery_time=RangeExpr.ALL
+    ))
+
+    self.import_kicad(self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
+                      conversions={
+                        'pwr_in': VoltageSink(
+                          current_draw=self.pwr_out.link().current_drawn,
+                          voltage_limits=RangeExpr.ALL,
+                        ),
+                        'pwr_out': VoltageSource(
+                          voltage_out=self.pwr_in.link().voltage,
+                          current_limits=RangeExpr.ALL,
+                        ),
+                        'control': DigitalSink(),  # TODO more modeling here?
+                        'gnd': Ground(),
+                        'btn_out': DigitalSingleSource(
+                          voltage_out=self.gnd.link().voltage-self.diode_drop.upper(),
+                          output_thresholds=(self.gnd.link().voltage.upper(), float('inf')),
+                          low_signal_driver=True
+                        ),
+                        'btn_in': DigitalSingleSource(
+                          voltage_out=self.gnd.link().voltage,
+                          output_thresholds=(self.gnd.link().voltage.upper(), float('inf')),
+                          low_signal_driver=True
+                        )
+                      })
+
+class SoftPowerSwitch(SoftPowerGate):
+  """A soft power switch that inherits from SoftPowerGate, with an additional
+  button for manual power control.
+  """
+  def contents(self):
+    self.btn = self.Block(Switch(voltage=0*Volt(tol=0)))  # TODO - actually model switch voltage
+    super().contents()
```

Comment:
This inevitably causes an error since the btn_in is no longer connected or needed here. 

- [ ] Q is there a way to remove the port in the parent?

Reply:
resolved with composition ver 
18eef59